Framework/Algorithms/src/ModeratorTzeroLinear.cpp

```diff
@@ -212,57 +215,35 @@ void ModeratorTzeroLinear::execEvent() {
 } // end of void ModeratorTzeroLinear::execEvent()
 
 // calculate time from sample to detector
-void ModeratorTzeroLinear::calculateTfLi(MatrixWorkspace_const_sptr inputWS,
+void ModeratorTzeroLinear::calculateTfLi(const SpectrumInfo &spectrumInfo,
                                          size_t i, double &t_f, double &L_i) {
   static const double convFact = 1.0e-6 * sqrt(2 * PhysicalConstants::meV /
                                                PhysicalConstants::NeutronMass);
   static const double TfError = -1.0; // signal error when calculating final
                                       // time
-  // Get detector position
-  IDetector_const_sptr det;
-  try {
-    det = inputWS->getDetector(i);
-  } catch (Exception::NotFoundError &) {
+
+  if (!spectrumInfo.hasDetectors(i)) {
     t_f = TfError;
     return;
   }
 
-  if (det->isMonitor()) {
-    L_i = m_instrument->getSource()->getDistance(*det);
+  if (spectrumInfo.isMonitor(i)) {
+    L_i = spectrumInfo.sourcePosition().distance(spectrumInfo.position(i));
```

Comment:
Well, `SpectrumInfo` defines L1+L2 = distance from source, and thus L2 is negative for monitors. That is what most algorithms use (and that is why `SpectrumInfo` does it). Here the algorithms wants something non-standard, so it is more explicit.